### PR TITLE
Forking UnixLinkerTool into AndroidLinkerTool.

### PR DESCRIPTION
**Pull request by benvanik**
There's some common things between them but that's mostly incidental at
this point. Once we have some more platforms working we can see if there
are attributes/etc we can apply everywhere (ensuring they're ignored when
not used on a particular tool/etc). For now, this stuff is whack-a-mole
and having them shared makes it harder to iterate on any one target.

With this our .so's for android end up with no imports and strictly the
functions we generate - an unstripped simple_mul is <500 bytes (most of
which is the source path names and junk in .comment).

Progress on #4606 - needs testing, script updates, and checks that things
like tracy still have symbols.

### CODE REVIEW

**Comment by benvanik** [comment]
Failing on android with UNAVAILABLE: https://buildkite.com/iree/iree-android-arm64-v8a/builds/3558
no clue why

File: build_tools/cmake/iree_check_test.cmake
Comment (asaadaldien) [inline build_tools/cmake/iree_check_test.cmake:82]: This is `os_kernel-ABI` it should be `linux-android${ABI_VERSION}`

**Review by asaadaldien — APPROVED** [review]
Nice! thanks for refactoring this @benvanik.

**Comment by asaadaldien** [comment]
Looking at the failure.

We need to statically link `libc` thats why I used to have `aarch64-linux-android30-clang++`  with `-static-libstdc++` 

```
/tmpfs/src/github/iree/iree/test/e2e/linalg_tensor_ops/add.mlir:14: undefined reference to `malloc'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
Linking failed; command line returned exit code 256:

/usr/src/android-ndk-r21d/toolchains/llvm/prebuilt/linux-x86_64/bin/aarch64-linux-android30-clang -nostdlib -static -lm -o /tmp/tensor_dispatch_0-54100e.so /tmp/tensor_dispatch_0-54100e.obj
```

**Comment by benvanik** [comment]
> Looking at the failure.
> 
> We need to statically link `libc` thats why I used to have `aarch64-linux-android30-clang++` with `-static-libstdc++`
> 
> ```
> /tmpfs/src/github/iree/iree/test/e2e/linalg_tensor_ops/add.mlir:14: undefined reference to `malloc'
> clang: error: linker command failed with exit code 1 (use -v to see invocation)
> Linking failed; command line returned exit code 256:
> 
> /usr/src/android-ndk-r21d/toolchains/llvm/prebuilt/linux-x86_64/bin/aarch64-linux-android30-clang -nostdlib -static -lm -o /tmp/tensor_dispatch_0-54100e.so /tmp/tensor_dispatch_0-54100e.obj
> ```

That was just a hack to silence the error and not a fix - we don't need to link libc - we shouldn't be emitting any code that uses anything from it - like malloc :)
We need to *not emit malloc*.

**Comment by ScottTodd** [comment]
This also helps with https://github.com/google/iree/issues/4717 (where most of the difficulty is coming from)

File: build_tools/mako/compile_android_modules.sh
Comment (ScottTodd) [inline build_tools/mako/compile_android_modules.sh:53]: triggered a build to test this: https://buildkite.com/iree/iree-android-benchmark/builds/1821

**Review by ScottTodd — APPROVED** [review]
Let's wait until ~Monday to merge? I'll be build cop then and want to sort out a few things downstream.

**Comment by benvanik** [comment]
> Let's wait until ~Monday to merge? I'll be build cop then and want to sort out a few things downstream.

SGTM

**Comment by ScottTodd** [comment]
Can we merge this now? I had some follow-up tasks to work through, but those can be made after this initial set of work lands.

**Comment by benvanik** [comment]
I'm fine with merging - I was waiting for your OK :)

**Comment by ScottTodd** [comment]
> I'm fine with merging - I was waiting for your OK :)

👍 I'm back around now to handle any follow-up tasks (or fallout from infra failures this may cause :P).

**Comment by ScottTodd** [comment]
> Hmm, this seems to break building on macOS:
> 
> `
> [2505/2862] Generating check_llvm-aot-conv_img2col_conv.mlir_module.vmfb
> FAILED: iree/test/e2e/llvm_specific/check_llvm-aot-conv_img2col_conv.mlir_module.vmfb
> cd /Users/xx/iree-build/iree/test/e2e/llvm_specific && /Users/xx/iree-build/iree/tools/iree-translate -iree-mlir-to-vm-bytecode-module --iree-hal-target-backends=dylib-llvm-aot -iree-codegen-linalg-to-llvm-conv-img2col-conversion=true /Users/xx/iree/iree/test/e2e/llvm_specific/conv.mlir -o check_llvm-aot-conv_img2col_conv.mlir_module.vmfb
> ld: unknown option: -nostdlib
> Linking failed; command line returned exit code 256:
> 
> /usr/bin/ld -nostdlib -shared -undefined suppress -o /var/folders/rp/c0_pxspj11g3dzxkc_qfxvdr0000gn/T/conv_linked_llvm_aot-8eec4a.so -dylib -flat_namespace /var/folders/rp/c0_pxspj11g3dzxkc_qfxvdr0000gn/T/conv_linked_llvm_aot-8eec4a.obj
> `

Oh, interesting. I also don't see the `nostdlib` option for `ld` on macOS (`man ld | grep nostdlib`), but I do see it on Linux. Thoughts, @benvanik ? As a workaround for now, we could wrap

https://github.com/google/iree/blob/39980126755bd0c5717a7c0d77c7366231a123ae/iree/compiler/Dialect/HAL/Target/LLVM/internal/UnixLinkerTool.cpp#L102-L104

in `#if defined(IREE_PLATFORM_APPLE)`...

**Comment by benvanik** [comment]
no no no no never add another #ifdef :)
(and definitely no deps on non-iree/compiler/ files)

you can query the target triple with `if (targetTriple.isOSDarwin()) ...` and change behavior on that, or the host triple (doing the compilation) with `auto hostTriple = Triple(llvm::sys::getProcessTriple()); if (hostTriple.isOSDarwin()) ...`.

Someone with a mac should figure out what the corresponding equivalent of nostdlib is and use that.

**Comment by benvanik** [comment]
(and, if things get weird, then add a DarwinLinkerTool just like AndroidLinkerTool)

**Comment by ScottTodd** [comment]
I found a discussion related to `nostdlib` on macOS here: https://reviews.llvm.org/D58320. Do we have someone familiar with linking set up already on a mac? I'm getting a new machine next week so I'd rather get set up on the new machine instead of spending that time now on an old machine :P

**Comment by asaadaldien** [comment]
> no no no no never add another #ifdef :)
> (and definitely no deps on non-iree/compiler/ files)
> 
> you can query the target triple with `if (targetTriple.isOSDarwin()) ...` and change behavior on that, or the host triple (doing the compilation) with `auto hostTriple = Triple(llvm::sys::getProcessTriple()); if (hostTriple.isOSDarwin()) ...`.
> 
> Someone with a mac should figure out what the corresponding equivalent of nostdlib is and use that.

We can also duplicate the list and explicitly `getFlags(targetTriple)` instead of concatenation.

**Pull request merged**
